### PR TITLE
142 jenkins file rap

### PR DESCRIPTION
**Pull request by RoddyMacrae**
## Pull Request submission 

*Insert detailed bullet points about your changes here!*

- Created a Jenkinsfile in the top-level directory to be included. This allows Jenkins to automatically pick up the branch and run the pipeline.
- Jenkins uses virtualenv and requirements.txt to run and test the pipeline. At the moment, to help in branches that don't have a requirements.txt and only a environment.yml, a convertor script has bene created. 
- Convertor script works as intended; however, Jenkins needs specific versions of certain packages to run and unless specified it will use the most up-to-date versions it can find, which in turn break. The requirements.txt file has the specified versions in there, which are:
    - coverage==4.5.4
    - pre-commit==2.17.0

- To allow the branch to be pushed, the .gitignore file had '!requirements.txt' added, and the pre-commit-config.yaml had 'txt' removed from restricted-filenames hook (on line 69 at time of writting). If these are not added/removed, the requirements.txt file gets ignored in the git push, or it gets blocked entirely.
- The pipelines can be viewed on Jenkins by logging into Jenkins, navigating into the ESG folder then resdev. Currently there are two branches, 142 and 36. 36 is a legacy pipeline that should be ignore. 142 is this ticket's branch. You should see any branch you include the aforementioned Jenkinsfile in (with requirements.txt set up as instructed using the convertor script) in this area on Jenkins. 
- Clicking on the branch in Jenkins - ESG/resdev you will see the pipeline stage view. If it is currently running you will see some stages greyed out. Successful stages will return green, unsuccesful ones red.
- Hovering over the different stages you should see a logs box pop up. If you click this you can view the logs from when the stage was run. 
- Another view which in my opinon is nicer to navigate/look at is the Blue Ocean view. You can open it on the left-hand side when in ESG/resdev.
- Blue Ocean will take you to the resdev overview and shows the branches that have been picked up and run through Jenkins. It gives a good overview of the health of the pipeline and the current run, helping to pinpoint the location of any instability or error. 
    - You can view the Blue Ocean dashboard page here: https://www.jenkins.io/doc/book/blueocean/dashboard/

- Provided you have your environement.yml (and therefore requirement.txt) set up to include all necessary packages, and have all unit tests in the ~/research-and-development/tests/ directory of your branch, Jenkins should run as intended and provide a report at the end to view either via the logs pop-ups or Blue Ocean.

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*

1. Run convertor script, yml_to_requirements_convert.py, if requirements.txt does not already exist/if it does not match the environment.yml file currently in use.
2. Make sure to specify the version numbers of the following:
    - coverage==4.5.4
    - pre-commit==2.17.0
3. If including new unit tests, make sure they are in the 'tests/' directory and names with the prefix test_XXXX.py
4. Make sure the file 'Jenkinsfile' is included in the top directory of your branch (~/research-and-development/)
5. Commit and push to your branch in order for Jenkins to pick up the branch via the Jenkinsfile
6. Check on Jenkins the status of the pipeline. It will likely take several minutes to be picked up depending on how busy the queue system is on Jenkins. After that, depending on the requirements and unit testing, it may take several minutes more to run. The current suggestion is to get up and stretch/get a coffee while waiting :)

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

N/A

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes RDRP-142: https://jira.ons.gov.uk/browse/RDRP-142

#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [x] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [x] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script) 
- [x] **Clean Code**
    - [x] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [x] Code passess flake8 linting check
    - [x] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions 
    - [ ] The docstring details data types
- [x] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [x] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

**Review by jwestw — CHANGES_REQUESTED** [review]
I have looked through your code and left comments. 
From what I can work out the Jenkins file looks good. I have checked Jenkins itself and triggered a workflow which seems to run fine. I am not quite sure how to see if the package is built correctly and where it is on Artifactory. 
The Python code you've written is a bit hard to follow as it has nested if statements in the for loop and there are tidier ways to do what you've done here. Creating functions should be your go to as they are testable and the docstring explains everything. I haven't written the whole solution, but I have created quite a bit of usable code for you in the `scripting_env_converter.ipynb` file that I have pushed under the notebooks folder. 

Also, a few questions for you:
- is the place that you've put `yml_to_requirements_convert.py` a suitable location?
- do we want to always run this script, or only when there is no `requirements.txt` present? 
- if this conversion is part of what's needed for the jenkins workflow, then can we make it part of the workflow? i.e. Can it be automated?
- are there any edits you'd suggest to the `environment.yml`?

File: yml_to_requirements_convert.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import yaml
+
+# yaml = ruamel.yaml.YAML()
+data = yaml.safe_load(open("/home/cdsw/research-and-development/environment.yml"))
```
[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply yml_to_requirements_convert.py]: Added

[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply yml_to_requirements_convert.py]: Created a more meaningful name in the context "yml_env"

File: yml_to_requirements_convert.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        if "=" in dep:
+            package, package_version = dep.split("=")
+            if package == "python":
+                requirements.append(package + "==3.6.2")
+            else:
+                requirements.append(package + "==" + package_version)
+        else:
+            requirements.append(dep)
+    elif isinstance(dep, dict):
+        for preq in dep.get("pip", []):
```
[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply yml_to_requirements_convert.py]: Created several functions to do what this does and provided further checks.

File: yml_to_requirements_convert.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            requirements.append(preq)
+    print(requirements)
+
+with open("/home/cdsw/research-and-development/requirements.txt", "w") as fp:
+    for requirement in requirements:
+        print(requirement, file=fp)
```
[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply yml_to_requirements_convert.py]: No not always. Written function to determine whether it needs updating.

File: yml_to_requirements_convert.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            if package == "python":
+                requirements.append(package + "==3.6.2")
+            else:
+                requirements.append(package + "==" + package_version)
+        else:
+            requirements.append(dep)
+    elif isinstance(dep, dict):
+        for preq in dep.get("pip", []):
+            requirements.append(preq)
+    print(requirements)
```
[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply yml_to_requirements_convert.py]: Missed.

**Comment by jwestw** [comment]
@RoddyMacrae Please can you delete the scripts that convert the yaml to requirements txt. Going forward we have now agreed to just use the `requirements.txt`, so please make sure that it is included and the `environment.yaml` is deleted. (Also please delete `previous-environmnet.yaml`). Once those scripts are deleted and those two files committed this ticket should be done.

**Review by jwestw — APPROVED** [review]
The code looks good and it runs. As far as I can see Jenkins is working.
